feat: add `math/base/special/cinvf`

anandkaranubc opened the pull request:
Progresses https://github.com/stdlib-js/stdlib/issues/649

## Description

> What is the purpose of this pull request?

This pull request:

-  Adds the complete C implementation for `math/base/special/cinvf`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   Progresses https://github.com/stdlib-js/stdlib/issues/649

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
feat: add C implementation for `math/base/special/cinvf`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

anandkaranubc committed:
refactor: use f32 instead of float64ToFloat32

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
Merge branch 'develop' into feat/cinvf

stdlib-bot committed:
docs: update namespace table of contents

PR-URL: https://github.com/stdlib-js/stdlib/pull/7062
Reviewed-by: Athan Reines <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/cinvf/README.md:
```diff
@@ -0,0 +1,251 @@
+var uniform = require( '@stdlib/random/base/uniform' );
+var cinvf = require( '@stdlib/math/base/special/cinvf' );
+
+var z1;
```
@anandkaranubc Would you mind refactoring to use `filledarrayBy` and `logEachMap` here? This way we can avoid the manual for loop.

kgryte commented on lib/node_modules/@stdlib/math/base/special/cinvf/examples/index.js:
```diff
@@ -0,0 +1,34 @@
+var uniform = require( '@stdlib/random/base/uniform' );
+var cinvf = require( './../lib' );
+
+var z1;
```
Same comment as README.

kgryte commented on lib/node_modules/@stdlib/math/base/special/cinvf/lib/main.js:
```diff
@@ -0,0 +1,107 @@
+	re = realf( z );
+	im = imagf( z );
+	ab = maxf( absf(re), absf(im) );
+	s = f32( 1.0 );
```
@anandkaranubc Rather than repeatedly do `f32(0.5)` and `f32(1.0)`, can you move those to the parent scope as

```
var ONE = f32( 1.0 );
var HALF = f32( 0.5 );
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/cinvf/src/main.c:
```diff
@@ -0,0 +1,99 @@
+// MAIN //
+
+/**
+Computes the inverse of a single-precision complex floating-point number.
```
```suggestion
* Computes the inverse of a single-precision complex floating-point number.
```

kgryte committed:
docs: fix missing asterisk

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (changes requested):
Other than the examples and some small clean-up, this is looking good.

anandkaranubc committed:
Merge branch 'develop' into feat/cinvf

anandkaranubc committed:
Merge branch 'develop' into feat/cinvf

anandkaranubc committed:
chore: apply suggestions from reviewa

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/cinvf/examples/index.js:
```diff
@@ -0,0 +1,32 @@
+
+'use strict';
+
+var Complex64 = require( '@stdlib/complex/float64/ctor' );
+var uniform = require( '@stdlib/random/base/uniform' );
+var filledarrayBy = require( '@stdlib/array/filled-by' );
+var logEachMap = require( '@stdlib/console/log-each-map' );
+var cinvf = require( './../lib' );
+
+function clbk() {
+	return new Complex64( uniform( -50.0, 50.0 ), uniform( -50.0, 50.0 ) );
+}
+var arr = filledarrayBy( 100, 'complex64', clbk );
+
+logEachMap( '1.0 / (%s) = %s', arr, cinvf );
```
@kgryte Does it look the way you intended to see it?

kgryte replied:
```suggestion
var Complex64Array = require( '@stdlib/array/complex64' );
var uniform = require( '@stdlib/random/array/uniform' );
var logEachMap = require( '@stdlib/console/log-each-map' );
var cinvf = require( './../lib' );

// Create an array of random numbers:
var arr = new Complex64Array( uniform( 200, -100.0, 100.0 ) );

// Compute the inverse of each number in the array:
logEachMap( '1.0 / (%s) = %s', arr, cinvf );
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/cinvf/README.md:
````diff
@@ -0,0 +1,249 @@
+<!-- eslint no-undef: "error" -->
+
+```javascript
+var Complex64 = require( '@stdlib/complex/float64/ctor' );
+var uniform = require( '@stdlib/random/base/uniform' );
+var filledarrayBy = require( '@stdlib/array/filled-by' );
+var logEachMap = require( '@stdlib/console/log-each-map' );
+var cinvf = require( '@stdlib/math/base/special/cinvf' );
+
+function clbk() {
+    return new Complex64( uniform( -50.0, 50.0 ), uniform( -50.0, 50.0 ) );
+}
+var arr = filledarrayBy( 100, 'complex64', clbk );
+
+logEachMap( '1.0 / (%s) = %s', arr, cinvf );
````
```suggestion
var Complex64Array = require( '@stdlib/array/complex64' );
var uniform = require( '@stdlib/random/array/uniform' );
var logEachMap = require( '@stdlib/console/log-each-map' );
var cinvf = require( '@stdlib/math/base/special/cinvf' );

// Create an array of random numbers:
var arr = new Complex64Array( uniform( 200, -100.0, 100.0 ) );

// Compute the inverse of each number in the array:
logEachMap( '1.0 / (%s) = %s', arr, cinvf );
```

anandkaranubc replied:
Much better... 😅

kgryte committed:
docs: update example

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: update example

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/cinvf/lib/main.js:
```diff
@@ -0,0 +1,109 @@
+
+// VARIABLES //
+
+var ONE = f32( 1.0 );
+var HALF = f32( 0.5 );
+var LARGE_THRESHOLD = f32( FLOAT32_BIGGEST * f32( 0.5 ) );
+var SMALL_THRESHOLD = f32( FLOAT32_SMALLEST * f32( f32(2.0)/EPS ) );
+var RECIP_EPS_SQR = f32( f32(2.0) / f32(EPS*EPS) );
```
```suggestion
var ONE = f32( 1.0 );
var TWO = f32( 2.0 );
var HALF = f32( 0.5 );
var LARGE_THRESHOLD = f32( FLOAT32_BIGGEST * HALF );
var SMALL_THRESHOLD = f32( FLOAT32_SMALLEST * f32( TWO/EPS ) );
var RECIP_EPS_SQR = f32( TWO / f32(EPS*EPS) );
```

kgryte committed:
refactor: use variable

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
